### DISCUSSION TITLE
Bring `bip322` into the `rust-bitcoin` org

### DISCUSSION THREAD

**Discussion by raphjaph** [rust-bitcoin/rust-bitcoin#3263] [General]
I'd like to request that the [bip322](https://github.com/raphjaph/bip322) be brought into the [rust-bitcoin](https://github.com/rust-bitcoin) org. 

- It has some basic CI
- I'm unsure if it adheres to `rust-bitcoin` org coding style and standards

I've implemented parts of the [BIP-322](https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki) specification. At the moment the crate supports signing and verifying messages for single-sig P2TR, P2WPKH and P2SH-P2WPKH addresses in the simple and full formats. I also plan to implement Proof-of-Funds and maybe multisig and legacy (BIP-137) support. 

The API is still very rough and I'd love to get feedback if I've implemented the cryptography correctly. Thanks!

**Comment by Kixunil** [comment]
Funny, past few days I've been working on something that uses generic signing, though legacy will do fine for now. IIUC Trezor doesn't support anything else.

Sounds cool to me though I suspect some people might be ideologically opposed to it. (I hope not.)

As a first check, you need to find out if it compiles with Rust 1.63 and if not modify it so it would.

**Reply by apoelstra** [reply]
> I suspect some people might be ideologically opposed to it

I don't think anyone in this project is ideologically opposed to it.

On the bips repo PR you can find a lot of really bizarre philosophical talk about what a signature "is" and "means" and some people are opposed on those grounds. But I'm not.

**Reply by raphjaph** [reply]
Thanks for the feedback! 

I've made it compile with Rust 1.63 in [this PR](https://github.com/raphjaph/bip322/pull/34). Should this check be added to the CI as well?

**Reply by apoelstra** [reply]
@raphjaph yes please!

If you are feeling really motivated, you could look at how our CI is setup here or in https://github.com/rust-bitcoin/rust-secp256k1/pull/699 (which is a little simpler) and copy that. We have a repo rust-bitcoin-maintainer-tools which provides scripts for all of our CI checks, though it's a little bit of a chore to set up.

**Reply by raphjaph** [reply]
Will check it out! I'm travelling at the moment but should be able to find some time the next couple days.

**Reply by storopoli** [reply]
If you have any doubts or need help don't hesitate in reaching out.
I am also excited for BIP 322 in `rust-bitcoin`.

**Reply by raphjaph** [reply]
I've added some of the CI in [this](https://github.com/raphjaph/bip322/pull/37) PR but removed a lot of things and unsure what exactly they do and if I need them.

- Do I need ASAN and a `sanitizer.sh` script?
- Not using `No-Std`
- I don't have any `EXAMPLES`, unsure what I should put here.
- No `minmal` or `recent` `Cargo.toml` files
- No WASM tests only check that it builds

I've also opened a small [PR](https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools/pull/15) fixing some typos in the `maintainer-tools` README.

Let me know how best to proceed.

**Reply by Kixunil** [reply]
* ASAN is not needed if the crate doesn't use `unsafe`.
* `no_std` would be really nice to have and there should be no reason why it wouldn't be possible but I think it's fine to not do it at first
* Examples would be nice but mainly they should be in docs. Depends on how complicated the API is.
* You have dependencies, so `minimal` and `recent` locks are highly recommended
* WASM is also nice to have and should be possible but I guess also OK to not do it (fully) at first.

**Reply by raphjaph** [reply]
Ok thanks, will add those things.

**Reply by apoelstra** [reply]
To be clear -- `no_std` support would be awesome, but you don't want a `no-std` *feature* if you can at-all avoid it. (Though I suspect you'll need one that just passes through to the `no-std` feature in `bitcoin`..)

But agreed it's fine not to have any nostd support at first. Ditto WASM.

And yes, please have `minimal` and `recent` lockfiles. Without them, every PR that touches Cargo.toml will require manual steps for me to review.

**Reply by tcharding** [reply]
You can just use any lock file that works and duplicate it as `Cargo-recent.lock` and `Cargo-minimal.lock`, then as time goes by we can update `Cargo-recent.lock`.

**Reply by raphjaph** [reply]
Ok I've incorporated the changes here: https://github.com/raphjaph/bip322/pull/37

**Reply by Kixunil** [reply]
@tcharding as the name `minimal` suggests, it should have the lowest possible versions (as restricted by `Cargo.toml`). So it should be produced by the `-Zminimal-versions` flag

**Reply by tcharding** [reply]
Of course, but you may find yourself spending a bunch of time pinning things and trying to get the build to work. I very soon get frustrated with that so I was just giving you a heads up that its acceptable for them to be the same.

**Reply by Kixunil** [reply]
If spending that time is undesirable then setting the same versions in `Cargo.toml` is appropriate.

**Reply by apoelstra** [reply]
I'd rather have no `Cargo-minimal.lock` then change our Cargo.toml to require versions we don't actually require.

It's bad enough that our dependencies fail to specify their dependencies properly (which is why `-Zminimal-versions` is such a pain to get working). We shouldn't make our own project worse because of it.

**Reply by Kixunil** [reply]
I'm operating on assumption that the crate already doesn't know which are the real versions it supports, so `Cargo.toml` is already wrong except in worse way - it reports that a version works when it doesn't as opposed to claiming that it doesn't but it does.

But good point that dependencies may create additional work.

**Reply by raphjaph** [reply]
When I run the `-Zminimal-versions` I run into the problems stated here: https://github.com/raphjaph/bip322/pull/37#issue-2513193963.

Unsure how to proceed. I have a `Cargo-recent.lock` and a `Cargo-minimal.lock` but they are not generated with the  `-Zminimal-versions`.

**Reply by yancyribbens** [reply]
Your crate built fine for me with `cargo +nightly build -Zminimal-versions`

**Reply by raphjaph** [reply]
Thanks, my Rust installation was borked somehow. It builds correctly now.

Can I merge that PR or is there anything else I should add?

**Reply by raphjaph** [reply]
I've merged https://github.com/raphjaph/bip322/pull/37. Let me know what to do to facilitate the next steps of review.

**Reply by apoelstra** [reply]
IMHO it's good to come in.

I'm not sure how to do this exactly -- in the past we've had repos like this which were owned by people who already had the permission to create repos in the org, and in that case Github gives you a nice "move repository" button.

I **think** you would need to give me or @tcharding admin rights on the repo, and then one of us can do the merge.

In any case -- ACK from me. Will wait for @tcharding to also ACK.

**Reply by raphjaph** [reply]
I've invited you as a collaborator but can also transfer ownership to you or the `rust-bitcoin` org.

**Reply by apoelstra** [reply]
Thanks!

After @tcharding takes a look I'll do the transfer.

**Reply by apoelstra** [reply]
I accepted the invitation but I still don't see the "Settings" tab where I would move the repo.

If you say you can transfer it @raphjaph how about you try?

**Reply by raphjaph** [reply]
I couldn't transfer it to the `rust-bitcoin` org because of missing permissions but I have transferred it to you @apoelstra now. I think you should then be able to transfer to `rust-bitcoin` org.

**Reply by apoelstra** [reply]
Perfect, thanks! I received it and re-transferred to rust-bitcoin.

https://github.com/rust-bitcoin/bip322

**Comment by tcharding** [comment]
LGTM, would be nice to upgrade to `bitcoin 0.32.0` soon as possible to be in lockstep with the rest of the ecosystem. I can help out with that. (I struggled to stop my neck twitching from the `tab_spaces = 2` config option :)

IMO this is ready to be brought into the org, thanks for your work!

**Reply by raphjaph** [reply]
Awesome! I'm happy to change any code style to be in line with the rest of the repositories :)

**Comment by raphjaph** [comment]
You guys probably also need publishing access for crates.io. Let me know who I should add there @apoelstra @tcharding.

If possible I'd also like to retain admin privileges on the transferred repo so I can cut releases and close PRs and Issues.

**Reply by apoelstra** [reply]
I would be fine leaving publishing access in your hands.

If you are the only publisher then yes, please add one of us to increase the bus factor. But (for my part) I would never use this privilege.

**Reply by tcharding** [reply]
Just adding @apoelstra on crates.io is best IMO. As for repo privileges, after moving `rust-psbt-v0` recently I still had all admin privileges and I don't have special org wide privileges. Conceptually you retain all privileges, rights, and accountability after the move - whether or not that works out of the box I don't know. @apoelstra is in charge of org wide settings, and has org wide privileges. It is totally still your crate, neither @apoelstra nor I will do stuff unless you want us to (I personally won't have any special privileges in the crate).

**Reply by raphjaph** [reply]
Ok sounds good, thanks for the clarification.
